### PR TITLE
Add kerberos support

### PR DESCRIPTION
**Pull request by rubenvp8510**
Signed-off-by: Ruben Vargas <ruben.vp8510@gmail.com>

This is a pure go solution.

### CODE REVIEW

**Comment by rubenvp8510** [comment]
I just signed the CLA.

**Comment by mieczkowski** [comment]
Hi. I've just created my solution for kerberos auth, but I will be happy to test your solution (native is better than libsasl2 <3 ).

File: gssapi_kerberos.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+	kt, _ := keytab.Load(krbAuth.config.KeyTabPath)
+	cfg, _ := krbconfig.Load(krbAuth.config.KerberosConfigPath)
+	krbAuth.client = krbclient.NewClientWithKeytab(krbAuth.config.Username, krbAuth.config.Realm, kt, cfg)
+	return nil
+}
+
+func (krbAuth *GSSAPIKerberosAuth) Authorize(broker *Broker) error {
+
+	krbAuth.createKerberosClient()
+	ticket, encryptionKey, err := krbAuth.client.GetServiceTicket(krbAuth.config.SPN)
```
Comment (mieczkowski) [inline gssapi_kerberos.go]: If we have kafka cluster then we have dynamic SPN, for example:

```Go
krbAuth.client.GetServiceTicket(krbAuth.config.SPN + "/" + strings.Split(broker.Addr(), ":")[0])
```

Client should be able to configure this behavior in some way.

Comment (rubenvp8510) [reply gssapi_kerberos.go]: Yes, that is true! Nice catch, Thanks!

**Comment by rubenvp8510** [comment]
@mieczkowski WDYT?

File: gssapi_kerberos.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+/*
+*
+* Construct Kerberos AP_REQ package, conforming to RFC-4120
+* https://tools.ietf.org/html/rfc4120#page-84
+*
+ */
+func (krbAuth *GSSAPIKerberosAuth) createKrb5Token(client *krb5client.Client, ticket messages.Ticket, sessionKey types.EncryptionKey) []byte {
+	var GSSAPIFlags = []int{gssapi.ContextFlagInteg, gssapi.ContextFlagConf}
+	auth, _ := types.NewAuthenticator(client.Credentials.Domain(), client.Credentials.CName())
```
Comment (mieczkowski) [inline gssapi_kerberos.go]: Missing error check

File: gssapi_kerberos.go [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+	auth.Cksum = types.Checksum{
+		CksumType: chksumtype.GSSAPI,
+		Checksum:  krbAuth.newAuthenticatorChecksum(GSSAPIFlags),
+	}
+	APReq, _ := messages.NewAPReq(
```
Comment (mieczkowski) [inline gssapi_kerberos.go]: Missing error check

File: gssapi_kerberos.go [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+		ticket,
+		sessionKey,
+		auth,
+	)
+	aprBytes, _ := hex.DecodeString(TOK_ID_KRB_AP_REQ)
```
Comment (mieczkowski) [inline gssapi_kerberos.go]: Missing error check

File: gssapi_kerberos.go [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+	tb, _ := APReq.Marshal()
```
Comment (mieczkowski) [inline gssapi_kerberos.go]: Missing error check

File: gssapi_kerberos.go [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+*	Append the GSS-API header to the payload, conforming to RFC-2743
+*	Section 3.1, Mechanism-Independent Token Format
+*
+*	https://tools.ietf.org/html/rfc2743#page-81
+*
+*	GSSAPIHeader + <specific mechanism payload>
+*
+ */
+func (krbAuth *GSSAPIKerberosAuth) appendGSSAPIHeader(payload []byte) []byte {
+	oidBytes, _ := asn1.Marshal(gssapi.OID(gssapi.OIDKRB5))
```
Comment (mieczkowski) [inline gssapi_kerberos.go]: Missing error check

File: gssapi_kerberos.go [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+	krbAuth.client = krb5client.NewClientWithKeytab(krbAuth.config.Username, krbAuth.config.Realm, kt, cfg)
+	return nil
+}
+
+/* This does the handshake for authorization */
+func (krbAuth *GSSAPIKerberosAuth) Authorize(broker *Broker) error {
+
+	err := krbAuth.createKerberosClient()
+	if err != nil {
+		Logger.Printf("Kerberos client error: %s", err)
```
Comment (mieczkowski) [inline gssapi_kerberos.go]: Missing `return err` (got panic on invalid config in next lines ;) )

Comment (rubenvp8510) [reply gssapi_kerberos.go]: Ooops.

File: gssapi_kerberos.go [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+func (krbAuth *GSSAPIKerberosAuth) createKerberosClient() error {
+	kt, err := keytab.Load(krbAuth.config.KeyTabPath)
+	if err != nil {
+		return err
+	}
+	cfg, err := krb5config.Load(krbAuth.config.KerberosConfigPath)
+	if err != nil {
+		return err
+	}
+	krbAuth.client = krb5client.NewClientWithKeytab(krbAuth.config.Username, krbAuth.config.Realm, kt, cfg)
```
[the remark this answers is not in the document]
Comment (rubenvp8510) [reply gssapi_kerberos.go]: I understand that could be semantically better but..

After thinking on this, I don't see the difference, now the `(c *Config) Validate()` happens on the Open method, same as kerberos client creation (where the keytab parsing happens), same place where authentication happens.

Also, If I follow the same logic  kerberos config validation, which happens when the kerberos client instance is created, I will need to move the kerberos client instantiation  to `(c *Config) Validate()`.

**Comment by rubenvp8510** [comment]
@mieczkowski @eapache @bai Could you review this one please?

Thanks

**Comment by rubenvp8510** [comment]
Any updates on this?

Thanks

**Comment by bai** [comment]
Thanks for your contribution!

**Comment by objectiser** [comment]
@bai Is there a release planned soon to include this feature?

**Comment by bai** [comment]
Yp, I'll cut a release in the next few days.

**Comment by objectiser** [comment]
@bai Thanks!

**Comment by objectiser** [comment]
Hi @bai, any update on when a release will be cut?

**Comment by shriram1993** [comment]
@rubenvp8510 I am trying to use kerberos authetication but getting error "**wrong Token ID. Expected 0504, was 6030"**. 
file name: **/gssapi_kerberos.go**
line no: 181
**if err := wrapTokenReq.Unmarshal(bytes, true); err != nil {return nil, err}**
Not getting any clue what i am doing wrong?
Please help.

**Comment by rubenvp8510** [comment]
@shriram1993 Could you please put information about your environment? version of kafka, how is it configured? how Kerberos is configured?

I see similar errors[1] on other libraries that relies on gokrb5 for handling Kerberos authentication messages, I think it could be something related to encryption.


1) https://github.com/colinmarc/hdfs/issues/145
Thanks.

**Comment by shriram1993** [comment]
Kafka version: 0.11.03
Kerberos Version: 1.15.1
Kerberos auth type: keytab file based
Kerberos server only supports **SASL_PLAINTEXT**.

The configuration set in **krb5.conf**
    ticket_lifetime =       1d
    renew_lifetime = 7d
    forwardable = true
    default_tgs_enctypes = **rc4-hmac**
    default_tkt_enctypes = **rc4-hmac**
    permitted_enctypes = **rc4-hmac**
    udp_preference_limit = 1
    kdc_timeout = 3000
    proxiable       =       true
    dns_lookup_realm = true
    dns_lookup_kdc = false
    allow_weak_crypto = false

**Comment by yzpnet** [comment]
How to support SASL_PLAINTEXT?

**Comment by shriram1993** [comment]
@rubenvp8510 I have the same question. My Kerberos server is expecting security protocol 'SASL_PLAINTEXT' but there is no way to set same in configuration. Please suggest.

**cfg := sarama.NewConfig()
cfg.Producer.Return.Successes = true
cfg.Metadata.Retry.Max = 0
cfg.Net.SASL.Enable = true
cfg.Net.SASL.Mechanism = sarama.SASLTypeGSSAPI
cfg.Net.SASL.GSSAPI.ServiceName = "kafka"
cfg.Net.SASL.GSSAPI.AuthType = sarama.KRB5_KEYTAB_AUTH
cfg.Net.SASL.GSSAPI.KeyTabPath = "<KEY TAB FILE PATH>"
cfg.Net.SASL.GSSAPI.KerberosConfigPath = "<KRB5 CONFIG PATH>"
cfg.Net.SASL.GSSAPI.Username = "<PRINCIPAL NAME>"
cfg.Net.SASL.GSSAPI.Realm = "<REALM>"**

**Comment by yzpnet** [comment]
@rubenvp8510 
I have the same question too. 
Kerberos auth need support SASL_PLAINTEXT.

**Comment by yzpnet** [comment]
SASL_PLAINTEXT

**Comment by rubenvp8510** [comment]
I'm going to start looking at how to implement SASL_PLAINTEXT support .@yzpnet  @shriram1993 Could you provide me your kafka configuration? That will help a lot.

I'm a little bit confused, In one comment I read "Kerberos server only supports SASL_PLAINTEXT." but as far as I know the problem is not in the communication with the KDC but with Kafka

I'm assuming that the implementation needs to support Kafka SASL_PLAINTEXT + Kerberos.  correct me if I misunderstood please.

Thank you!

**Comment by shriram1993** [comment]
**Kafka config:**
        properties.setProperty("bootstrap.servers", "kafka-hotels.example.com:9092");
        properties.setProperty("key.serializer.class", "kafka.serializer.StringEncoder");
        properties.setProperty("serializer.class", "kafka.serializer.DefaultEncoder");
        properties.setProperty("request.required.acks", "1");
        properties.setProperty("retries", "3");
        properties.setProperty("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");
        properties.setProperty("value.serializer", "org.apache.kafka.common.serialization.StringSerializer");
        properties.setProperty("compression.type", "snappy");
        properties.setProperty("request.timeout.ms", "300000");
        properties.setProperty("metadata.fetch.timeout.ms", "100000");
        properties.setProperty("security.protocol", "SASL_PLAINTEXT");
        properties.setProperty("sasl.kerberos.service.name", "kafka");
        System.setProperty("java.security.krb5.conf", "/etc/kerberos/conf/krb5.conf");
        System.setProperty("java.security.auth.login.config", "path of jaas-cache.conf");

**jaas-cache.conf:**
KafkaClient {
     com.sun.security.auth.module.Krb5LoginModule required
     useKeyTab=true
     keyTab="keytab.keytab"
     storeKey=true
     useTicketCache=false
     serviceName="kafka"
     principal="mirroruser@EXAMPLE.COM";
    };

**Krb5.conf:**
default_realm = EXAMPLE.COM
    clockskew       =       300
    ticket_lifetime =       1d
    renew_lifetime = 7d
    forwardable = true
    default_tgs_enctypes = rc4-hmac
    default_tkt_enctypes = rc4-hmac
    permitted_enctypes = rc4-hmac
    udp_preference_limit = 1
    kdc_timeout = 3000
    proxiable       =       true
    dns_lookup_realm = true
    dns_lookup_kdc = true
    allow_weak_crypto = true

**Comment by shriram1993** [comment]
> I'm a little bit confused, In one comment I read "Kerberos server only supports SASL_PLAINTEXT." but as far as I know the problem is not in the communication with the KDC but with Kafka 
> 
> I'm assuming that the implementation needs to support Kafka SASL_PLAINTEXT + Kerberos. correct me if I misunderstood please.
> 
> Thank you!

Yes!! You are right. The issue is not with KDC it is with Kafka.
Kerberos only support SASL_PLAINTEXT means we are using Kerberos + Kafka with this security protocol SASL_PLAINTEXT.

**Comment by rubenvp8510** [comment]
well, this should work with SASL_PLAINTEXT by default, I'm reviewing what would cause the error you are seeing.

**Comment by rubenvp8510** [comment]
Hi @shriram1993  @yzpnet 

After reviewing this in detail, I was managed to reproduce the issue, it seems like the actual implementation of Kerberos authentication does not support rc4-hmac.

The thing here is that the RFC-4121 (or at least the JAVA implementation) only support the use of aes128-cts-hmac-sha1-96 and  aes256-cts-hmac-sha1-96 as a method for encrypt and integrity check for tkt and tgs. For work with rc4-hmac we would need to implement RFC-1964.

So, I would say this should work if you configure your Active Directory[1] to support those encrypt types and your krb5.conf to use it (not sure but I think you need to regenerate your keytab files if you are using). You can also need to disable FAST negotiation, I'll send a PR  to expose that option for sarama.

Let me know if this works for you so I can close the corresponding issue. I'll test it and post my configs here.
 
1)https://blogs.msdn.microsoft.com/openspecification/2011/05/30/windows-configurations-for-kerberos-supported-encryption-type/

**Comment by rubenvp8510** [comment]
Hi again

Confirmed that the issue is with **rc4-hmac**

This is my configuration:

```
[logging]
    default = FILE:/var/log/krb5libs.log
    kdc = FILE:/var/log/krb5kdc.log
    admin_server = FILE:/var/log/kadmind.log

[libdefaults]
    default_realm = EXAMPLE.COM
    clockskew = 300
    ticket_lifetime = 1d
    renew_lifetime = 7d
    forwardable = true
    default_tgs_enctypes = aes256-cts-hmac-sha1-96
    default_tkt_enctypes = aes256-cts-hmac-sha1-96
    permitted_enctypes = aes256-cts-hmac-sha1-96
    udp_preference_limit = 1
    kdc_timeout = 3000
    proxiable = true
    dns_lookup_realm = true
    dns_lookup_kdc = true
    allow_weak_crypto = true

[realms]
EXAMPLE.COM = {
     kdc = 192.168.58.3
     admin_server = 192.168.58.3
}

[domain_realm]
 .kerberoswin.com = EXAMPLE.COM
  kerberoswin.com = EXAMPLE.COM

```

Followed the guide mentioned in my previous comment to configure AES encrypt types for my user accounts and regenerate keytabs for my services.

I regenerated my keytabs using this command:

` ktpass -princ kafka/kafka@EXAMPLE.COM -mapuser EXAMPLE0\kafka -crypto AES256-SHA1 -ptype KRB5_NT_PRINCIPAL -pass kfk_password -out C:\keytabs\kafka.ktab`

**Pull request merged**